src/System.Private.CoreLib/shared/System/Collections/Generic/Dictionary.cs

```diff
@@ -1153,6 +1166,22 @@ void IDictionary.Remove(object key)
             }
         }
 
+#if BIT64
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private ref int GetBucket(uint hashCode)
+        {
+            int[] buckets = _buckets!;
+            return ref buckets[HashHelpers.FastMod(hashCode, (uint)buckets.Length, _fastModMultiplier)];
+        }
+#else
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private ref int GetBucket(uint hashCode)
```

Comment:
Great! Both variations have the same signature, so we may leave under `#if` the index calculation only. And if we decide to use `Unsafe` to avoid the bounds check, it will be in one place only.

Reply:
Merged

Reply:
> we decide to use Unsafe to avoid the bounds check

Note that we want collections to be secure without buffer overruns even if the calling code has race conditions. It is not ok to skip the boundscheck for the 64-bit case.

And skipping the bounds check for the 32-bit has immaterial benefit (the bounds checks costs "nothing" compare to the modulo. If we really wanted this optimization, it should be done in the JIT.

Reply:
@jkotas If we cache the `_buckets` value, its `Length` cannot change, so no race condition.  Unfortunately, JIT has no knowledge that the `Length` is non-zero.

Reply:
Can combine the bucket array allocation, prime and multiplier all in the non-generic HashHelpers

Reply:
Mod zero throws; so the Jit doesn't need to worry about length being zero

Reply:
> Mod zero throws

You are right and I was confused by comments in #15472. So JIT might handle the modulo case, but not FastMod.